meds_interp/knn_raw_code.py

```diff
@@ -0,0 +1,429 @@
+        if weights is None:
+            weights = np.ones_like(neigh_ind)
+
+        all_rows = np.arange(n_queries)
+        # probabilities = []
+
+        classes = self.classes_
+        pred_labels = _y[neigh_ind]
+        proba = np.zeros((n_queries, classes.size))
+
+        # a simple ':' index doesn't work right
+        for i, idx in enumerate(pred_labels.T):  # loop is O(n_neighbors)
+            proba[all_rows, idx] += weights[:, i]
+
+        # normalize 'votes' into real [0,1] probabilities
+        normalizer = proba.sum(axis=1)[:, np.newaxis]
+        proba /= normalizer
+        return proba
+
+    def predict_proba(self, X):
+        """Estimates the posterior probabilities for sample in X.
+
+        Parameters
+        ----------
+        X : array of shape (n_samples, n_features)
+            The input data.
+
+        Returns
+        -------
+        preds_proba : array of shape (n_samples, n_classes)
+                          Probabilities estimates for each sample in X.
+        """
+        if self.weights == "uniform":  # uses closest neighbors and treats all labels equally
+            idx = self.kneighbors(X, self.n_neighbors, return_distance=False)
+            return self.get_uniform_proba(idx)
+        else:  # weights neighbor's labels by inverse distance to the neighbor
+            assert self.weights == "distance"
+            neigh_dist, neigh_ind = self.kneighbors(X, self.n_neighbors, return_distance=True)
+            return self.get_distance_proba(neigh_dist, neigh_ind)
+
+    def concat_data(self, X: list[np.array] | pl.DataFrame):
+        embeddings = []
+        if isinstance(X, pl.DataFrame):
+            for modality in self.modalities:
+                embed_array = np.asarray(X[modality].to_list())
+                embeddings.append(embed_array)
+        else:
+            embeddings = X
+        return np.concatenate(embeddings, axis=1)
+
+    def fit_preprocess(self, X: pl.DataFrame):
+        if self.preprocess == Preprocess_Type.NORM_AFTER_CONCAT:
+            self.scaler = Normalizer()
+            self.scaler.fit(X)
+        elif self.preprocess == Preprocess_Type.NORM_SEPERATLY:
+            self.scalers = []
+            assert X.shape[1] == self.d * 2
+            assert self.modalities.length() == self.modality_weights.length()
+            for i in range(self.modalities.length()):
+                embed_array = np.asarray(X[self.modalities[i]].to_list())
+                scaler = Normalizer()
+                scaler.fit(embed_array)
+                self.scalers.append(scaler * self.modality_weights[i])
+            # for modality in self.modalities:
+            #     embed_array = np.asarray(X[modality].to_list())
+            #     scaler = Normalizer()
+            #     scaler.fit(embed_array)
+            #     self.scalers.append(scaler)
+        else:
+            assert self.preprocess == Preprocess_Type.NONE
+
+    def transform_preprocess(self, X):
```

Comment:
We should add type definitions to the functions^. For example X should be `X: np.array` I think.